mlir/lib/Dialect/MIOpen/Transforms/ThreadwiseToStdlib.cpp

```diff
@@ -906,116 +907,81 @@ struct ThreadwiseGemmRewritePattern
 
   LogicalResult matchAndRewrite(ThreadwiseGemmOp op,
                                 PatternRewriter &b) const override {
-    auto loc = op.getLoc();
+    Location loc = op.getLoc();
 
-    auto gemmA = op.matrixA();
-    auto gemmB = op.matrixB();
-    auto gemmC = op.matrixC();
-    auto dataType =
+    Value gemmA = op.matrixA();
+    Value gemmB = op.matrixB();
+    Value gemmC = op.matrixC();
+    Type dataType =
         gemmA.getType().template cast<MemRefType>().getElementType();
 
-    ArrayRef<int64_t> gemmAShape =
-        gemmA.getType().cast<MemRefType>().getShape();
-    ArrayRef<int64_t> gemmBShape =
-        gemmB.getType().cast<MemRefType>().getShape();
-
-    assert(gemmAShape.size() == gemmBShape.size());
-    assert((gemmAShape.size() == 3) || (gemmAShape.size() == 4));
-    if (gemmAShape.size() == 3) {
-      // non-KPack path.
-      auto loopG = b.create<AffineForOp>(loc, 0, gemmAShape[0]);
-      auto lbG = loopG.getBody();
-      b.setInsertionPointToStart(lbG);
-
-      auto loopK = b.create<AffineForOp>(loc, 0, gemmAShape[1]);
-      auto lbK = loopK.getBody();
-      b.setInsertionPointToStart(lbK);
-
-      auto loopM = b.create<AffineForOp>(loopK.getLoc(), 0, gemmAShape[2]);
-      auto lbM = loopM.getBody();
-      b.setInsertionPointToStart(lbM);
-
-      auto loopN = b.create<AffineForOp>(loc, 0, gemmBShape[2]);
-      auto lbN = loopN.getBody();
-      b.setInsertionPointToStart(lbN);
-
-      SmallVector<Value, 3> memIndicesKM;
-      extractForInductionVars({loopG, loopK, loopM}, &memIndicesKM);
-      auto gemmAKM = b.create<AffineLoadOp>(loc, gemmA, memIndicesKM);
-
-      SmallVector<Value, 3> memIndicesKN;
-      extractForInductionVars({loopG, loopK, loopN}, &memIndicesKN);
-      auto gemmBKN = b.create<AffineLoadOp>(loc, gemmB, memIndicesKN);
-
+    int64_t g = op.gAttr().getInt();
+    int64_t k = op.kAttr().getInt();
+    int64_t m = op.mAttr().getInt();
+    int64_t n = op.nAttr().getInt();
+    int64_t kPack = op.kPackAttr().getInt();
+    LLVM_DEBUG(llvm::dbgs() << "Threadwise gemm:\n"
+                            << "g = " << g << "\n"
+                            << "k = " << k << "\n"
+                            << "m = " << m << "\n"
+                            << "n = " << n << "\n"
+                            << "kPack = " << kPack << "\n");
+    SmallVector<int64_t> dimensions = {g, k, m, n, kPack};
```

Comment:
I like it that you are using coordinate transfer for the inner k/m/n computation. However, for the inner most kpack dimension, will it get consolidated as a vectorized load/store/computation?

Reply:
Ok, so, with `kPack = 2`, the code will be 
```
c[0 * N + 0] = c [0 * N + 0] + a[(0 * M + 0) * kPack + 0] + b[(0 * N + 0) * kPack + 0]
c[0 * N + 0] = c[0 * N + 0] + a[(0 * M + 0) * kPack + 1] + b[(0 * N + 0) * kPack + 1]
c[0 * N + 1] = c[0 * N + 1] + a[(0 * M + 0) * kPack + 0] + b[(0 * N + 1) * kPack + 0]
...
```
And once this is all SSA-ified, that intermediate store/load pair will go away.